src/builtin.jq

```diff
@@ -6,6 +7,19 @@ def sort_by(f): _sort_by_impl(map([f]));
 def group_by(f): _group_by_impl(map([f]));
 def unique: group_by(.) | map(.[0]);
 def unique_by(f): group_by(f) | map(.[0]);
+# max(s) and min(s) use boxing technique for the sake of `input`:
+def max(s):
+  reduce (s|[.]) as $x (null;
```

Comment:
The version in my earlier comment didn't allocate an array for every `$x`.

Reply:
Yes, but setpath can slow it down, as it did in my tests.  Fixing `input` would be so much nicer!

Reply:
What do you mean "fixing `input` would be so much nicer"?

As for `setpath`... we're using it on a zero-or-one element array, so it should not allocate at all after the first time...  I'll check that out later.